File: app/src/main/java/com/foxek/simpletimer/ui/timer/TimerInteractor.java
package com.foxek.simpletimer.ui.timer;

import com.foxek.simpletimer.data.model.Interval;
import com.foxek.simpletimer.utils.AlarmHelper;
import com.foxek.simpletimer.utils.TimerHelper;
import com.foxek.simpletimer.data.database.TimerDatabase;


import java.util.ArrayList;

import javax.inject.Inject;

import io.reactivex.Flowable;
import io.reactivex.Observable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;

import static com.foxek.simpletimer.utils.Constants.EMPTY;
import static com.foxek.simpletimer.utils.Constants.PREPARE_INTERVAL;

public class TimerInteractor implements TimerContact.Interactor {

    private TimerDatabase database;
    private ArrayList<Integer> intervalList;
    private ArrayList<String> nameList;
    private TimerHelper timerHelper;
    private AlarmHelper alarmHelper;

    @Inject
    TimerInteractor(TimerDatabase database, AlarmHelper alarmHelper, TimerHelper timerHelper) {
        this.database = database;
        this.timerHelper = timerHelper;
        this.alarmHelper = alarmHelper;
    }

    public void deleteDependencies() {
        timerHelper.timerDelete();
    }

    @Override
    public String getIntervalName(int id) {
        return nameList.get(id);
    }

    @Override
    public Flowable<Integer> fetchIntervalList(int workoutId) {
        intervalList = new ArrayList<>();
        nameList = new ArrayList<>();
        intervalList.add(PREPARE_INTERVAL);
        nameList.add(EMPTY);
        return database.getIntervalDAO().getAll(workoutId)
                .map(intervals -> {
                    for (Interval interval : intervals) {
                        intervalList.add(interval.getWorkTime());
                        intervalList.add(interval.getRestTime());
                        nameList.add((interval.getName()));
                    }
                    return intervalList.size();
                });
    }

    @Override
    public Disposable getVolume(int id) {
        return database.getWorkoutDAO().getVolume(id)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(volume -> alarmHelper.setVolume(volume), throwable -> {
                });
    }

    @Override
    public void loadIntervalListToTimer() {
        timerHelper.loadIntervalList(intervalList);
    }

    @Override
    public void continueTimer() {
        timerHelper.timerRecreate();
    }

    @Override
    public void stopTimer() {
        timerHelper.timerDelete();
    }

    @Override
    public Observable<Integer> intervalFinishedCallback() {
        return timerHelper.onIntervalFinished();

    }

    @Override
    public Observable<Integer> tickCallback() {
        return timerHelper.onTimerTickHappened()
                .map(time -> {
                    if (((time > 500) && (time < 1000)) || ((time > 1500) && (time < 2000)) ||
                            ((time > 2500) && (time < 3000)))
                        indicateLastSeconds();
                    return (int) ((time / 1000) + 1);
                });
    }

    @Override
    public boolean getTimerState() {
        return timerHelper.getTimerState();
    }

    @Override
    public void setTimerState(boolean timerState) {
        timerHelper.setTimerState(timerState);
    }

    @Override
    public int getIntervalSize() {
        return intervalList.size();
    }

    @Override
    public void indicateEndOfWorkout() {
        alarmHelper.patternVibrate();
        alarmHelper.playFinalSound();
    }

    @Override
    public void indicateEndOfInterval() {
        alarmHelper.oneShotVibrate();
        alarmHelper.playLongSound();
    }

    @Override
    public void indicateLastSeconds() {
        alarmHelper.playSound();
    }
}

File: app/src/main/java/com/foxek/simpletimer/ui/timer/TimerInteractor.kt
package com.foxek.simpletimer.ui.timer

import com.foxek.simpletimer.data.database.TimerDatabase
import com.foxek.simpletimer.data.model.Interval
import com.foxek.simpletimer.data.model.Time
import com.foxek.simpletimer.utils.Constants.EMPTY
import com.foxek.simpletimer.utils.Constants.POST_TIME_TYPE
import com.foxek.simpletimer.utils.Constants.PREPARE_TIME
import com.foxek.simpletimer.utils.Constants.PREPARE_TIME_TYPE
import com.foxek.simpletimer.utils.Constants.REST_TIME_TYPE
import com.foxek.simpletimer.utils.Constants.WITH_REST_TYPE
import com.foxek.simpletimer.utils.Constants.WORK_TIME_TYPE
import com.foxek.simpletimer.data.timer.IntervalTimer
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.schedulers.Schedulers
import java.util.ArrayList
import javax.inject.Inject

class TimerInteractor @Inject constructor(
        private val database: TimerDatabase,
        private val timer: IntervalTimer
) : TimerContact.Interactor {

    private var times = ArrayList<Time>()

    override fun deleteDependencies() {
        timer.stop()
    }

    override fun fetchIntervalList(workoutId: Int): Disposable {
        return database.intervalDAO.getAll(workoutId)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .doOnNext { fetchTimeList(it) }
                .subscribe({ timer.prepare(times) }, {})
    }

    override fun getVolume(id: Int): Disposable {
        return database.workoutDAO.getVolume(id)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ volume -> timer.enableSound(volume) }, { })
    }

    override fun continueTimer() {
        timer.restart()
    }

    override fun stopTimer() {
        timer.stop()
    }

    override fun intervalFinishedCallback(): Observable<Time> {
        return timer.onIntervalFinished
    }

    override fun tickCallback(): Observable<Int> {
        return timer.onTimerTickHappened
    }

    override fun getTimerState(): IntervalTimer.State {
        return timer.state
    }

    private fun fetchTimeList(intervalList: List<Interval>) {

        times.add(Time(PREPARE_TIME, PREPARE_TIME_TYPE, 0, EMPTY))

        intervalList.forEachIndexed { idx, it ->
            times.add(Time(it.work, WORK_TIME_TYPE, idx + 1, it.name))

            if (it.type == WITH_REST_TYPE)
                times.add(Time(it.rest, REST_TIME_TYPE, idx + 1, it.name))
        }

        times.add(Time(0, POST_TIME_TYPE, times.lastIndex, null))
    }
}
